### DISCUSSION TITLE
JSON schema for unist/mdast

### DISCUSSION THREAD

**Discussion by chrisjsewell** [syntax-tree/unist#61] [Q&A] [answered]
Heya, Is there a "ratified" JSON schema for unist (and also mdast)?

**Comment by chrisjsewell** [comment]
Something along the lines of this is what I came up with?

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "description": "Syntactic unit in unist syntax trees.",
  "type": "object",
  "$defs": {
    "point": {
      "description": "A point represents one place in a source file.",
      "type": "object",
      "required": ["line", "column"],
      "properties": {
        "line": {
          "description": "The line number of the point.",
          "type": "integer",
          "minimum": 1
        },
        "column": {
          "description": "The column number of the point.",
          "type": "integer",
          "minimum": 1
        },
        "offset": {
          "description": "The offset of the point.",
          "type": "integer",
          "minimum": 0
        }
      }
    },
    "position": {
      "description": "A position represents the location of a node in a source file.",
      "type": "object",
      "required": ["start", "end"],
      "properties": {
        "start": {
          "description": "The start position of the position.",
          "$ref": "#/definitions/point"
        },
        "end": {
          "description": "The end position of the position.",
          "$ref": "#/definitions/point"
        },
        "indent": {
          "description": "The indentation of the position.",
          "type": "integer",
          "minimum": 0
        }
      }
    }
  },
  "required": ["type"],
  "properties": {
    "type": {
      "description": "The type of the node.",
      "type": "string"
    },
    "position": {
      "$ref": "#/definitions/position"
    },
    "data": {
      "description": "Information associated by the ecosystem with the node.",
      "type": "object"
    },
    "children": {
      "description": "The children of a Parent node.",
      "type": "array",
      "items": {
        "$ref": "#"
      }
    },
    "value": {
      "description": "The value of a Literal node."
    }
  }
}
```

**Comment by ChristianMurphy** [comment, marked as answer]
There are some assertion libraries that check that various syntax trees have expected structure:
* https://github.com/syntax-tree/unist-util-assert
* https://github.com/syntax-tree/hast-util-assert
* https://github.com/syntax-tree/mdast-util-assert
* https://github.com/syntax-tree/nlcst-test

There are also standards documents for each:

* https://github.com/syntax-tree/unist
* https://github.com/syntax-tree/mdast
* https://github.com/syntax-tree/hast
* https://github.com/syntax-tree/nlcst

I don't know if there an official schema 🤔 
/cc @remcohaszing I know you worked on JSON schemas for configuration files, did you also create some for the trees themselves?

The one you shared @chrisjsewell looks about right.

**Reply by chrisjsewell** [reply]
Yeh I want to use it for a Python application, so can't use any JS libraries 😄
I'm sure someone has done it before, because GitHub copilot autocompleted some of it lol

**Reply by remcohaszing** [reply]
> I'm sure someone has done it before, because GitHub copilot autocompleted some of it lol

Haha that’s one way to find out. :smile: 

---

I haven’t created them, but it should be fairly straightforward to create them based on specifications linked by @ChristianMurphy.

[JSON Schema Store](https://github.com/SchemaStore/schemastore) accepts pretty much any pull request adding a JSON schema. AFAIK those schemas are linked to a filename though.
